### ISSUE TITLE
IPv6 addresses not parsed correctly in /usr/mailcleaner/bin/dump_firewall.pl

### ISSUE THREAD

**Issue by goodomens42** [MailCleaner/MailCleaner#313]
The function get_external_rules in /usr/mailcleaner/bin/dump_firewall.pl does not parse IPv6 addresses correctly due to a wrong pattern.

Sugested patch:

*** dump_firewall.pl    2021-02-23 19:33:29.261450000 +0100
--- dump_firewall.pl.orig       2021-02-23 19:31:55.637450000 +0100
***************
*** 130,136 ****
       next if ($ref->{'port'} !~ /^\d+[\:\|]?\d*$/);
       next if ($ref->{'protocol'} !~ /^(TCP|UDP|ICMP)$/i);
       foreach my $ip (expand_host_string($ref->{'allowed_ip'})) {
!        if ($ip =~ m/^[a-fA-F0-9\.\:]{3,40}\/?\d*$/) {
             $rules{"host ".$ip.", service ".$ref->{'service'}} = [ $ref->{'port'}, $ref->{'protocol'}, $ip];
         }
       }
--- 130,136 ----
       next if ($ref->{'port'} !~ /^\d+[\:\|]?\d*$/);
       next if ($ref->{'protocol'} !~ /^(TCP|UDP|ICMP)$/i);
       foreach my $ip (expand_host_string($ref->{'allowed_ip'})) {
!        if ($ip =~ m/^[0-9\.\:]{3,40}\/?\d*$/) {
             $rules{"host ".$ip.", service ".$ref->{'service'}} = [ $ref->{'port'}, $ref->{'protocol'}, $ip];
         }
       }

**Comment by andersonvaz** [comment]
I have problem in IPV4 too

telnet mailc3.deploycloud.com.br 25
Trying 198.204.254.172...
Connected to mailc3.deploycloud.com.br.
Escape character is '^]'.
550 Administrative prohibition
Connection closed by foreign host.

**Comment by JohnMertz** [comment]
> The function get_external_rules in /usr/mailcleaner/bin/dump_firewall.pl does not parse IPv6 addresses correctly due to a wrong pattern.

@goodomens42 Thanks. It appear that this pattern is not new. Regardless, the expand_host_string function already validates IPs. It will also return hostnames, so I'm currently testing a commit that will just check for CIDR notation (all returned IP ranges are converted to CIDR, even /32 or /128): https://github.com/MailCleaner/MailCleaner/pull/314

It is bundled with a small change to the new hostlist expansion to push any invalid hostlist entries to a log file instead of STDERR.

> I have problem in IPV4 too
> 
> telnet mailc3.deploycloud.com.br 25
> Trying 198.204.254.172...
> Connected to mailc3.deploycloud.com.br.
> Escape character is '^]'.
> 550 Administrative prohibition
> Connection closed by foreign host.

@andersonvaz This may actually be related to the exim dumper which was reverted earlier today. A corrected commit is included in the pull request that I've referenced above, but a regular update should revert to the previous dump script until this request is merged. You can either run /root/Updater4MC/updater4mc.sh or pull the git directory (/usr/mailcleaner) and restart exim_stage1.

**Comment by JohnMertz** [comment]
#314 was merged this morning. You can force an update with /root/Updater4MC/updater4mc.sh to see if this resolves the issue.

**Comment by goodomens42** [comment]
@JohnMertz 
_> It appear that this pattern is not new_

You are right, first time I stumbled accross this problem was in early December, when we introduced IPv6 on our mail gateways.
I created my own patch (see above) which fixed the problem, but was too lazy/busy/whatever to actually report the bug.
The update published on 2021-02-22 brought back the problem as well as issue #310, so I reported both this time :)

_> #314 was merged this morning.
> You can force an update with /root/Updater4MC/updater4mc.sh to see if this resolves the issue._

Your patch seems to work fine, ip6tables -L -n lists all required rules after update and reboot.
Many thanks, good work !
